### PR TITLE
[ZKD] Sparse Property

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
`mdi` (or `zkd`) and `mdi-prefixed` now support `sparse: true`.

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp [hunk 1]
```diff
@@ -763,16 +788,32 @@ Result RocksDBZkdIndex::remove(transaction::Methods& trx,
...
+    keyValue = std::move(result.get());
+  }
 
   RocksDBKey rocksdbKey;
   uint64_t hash = 0;
   if (!isPrefixed()) {
     rocksdbKey.constructZkdIndexValue(objectId(), keyValue, documentId);
   } else {
-    auto prefixValues = extractAttributeValues(trx, _prefixFields, doc);
+    auto result = extractAttributeValues(trx, _prefixFields, doc, !_sparse);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBZkdIndex.cpp:810]: The new parameter `nullAllowed` is now set to true if `!_sparse`, i.e. we allow null values if the index is not sparse.
Shouldn't this be opposite, i.e. 
```suggestion
    auto result = extractAttributeValues(trx, _prefixFields, doc, _sparse);
```
?

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp [hunk 2]
```diff
@@ -691,6 +694,11 @@ auto extractAttributeValues(
     if (s.isNone()) {
       s = VPackSlice::nullSlice();
     }
+
+    if (s.isNull() && !nullAllowed) {
+      return {TRI_ERROR_ARANGO_DOCUMENT_KEY_MISSING};
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBZkdIndex.cpp:699]: Is this the best error code we can use here? In other cases we are using it to indicate that the `_key` attribute is missing.
This is not the case here. I am happy if we use this error code here too, but it should be a conscious decision.

Comment (maierlars) [reply arangod/RocksDBEngine/RocksDBZkdIndex.cpp:699]: The code is only used internally and never leaves the database. It's only used within this file. Happy to change it, I just don't want to add a new error code here. Maybe something more generic?

Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBZkdIndex.cpp:699]: If it is used only internally, I think it is fine as is. Would be useful to add a code comment somewhere to indicate that this is only an internal code to signal a certain condition.

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp (hunk 2 above)
Comment (maierlars) [reply arangod/RocksDBEngine/RocksDBZkdIndex.cpp:699]: I will add that comment in a follow up PR.

**Pull request merged**